Make functionalization `ViewMeta` serializable with pickle.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143712

Fix: #141974

This PR makes `ViewMeta` sequence, present in functional tensors,
serializable with pickle. In order to accomplish that, it makes
`ViewMeta` an abstract class with overridable `forward` and `reverse`
functions. In this context, each operation that once instanciated
`ViewMeta`, should now create a new specialized class that inherits from
`ViewMeta. Therefore, this PR also uses codegen for creating these
specializations.

In summary, these are the changes this PR introduces:

- `ViewMeta` is turned into an abstract class (see
  _FunctionalStorageImpl.cpp_). `forward` and `reverse` are pure virtual
  functions that need to be implemented. `to_out_index` should be
  implemented by operations that might return more than 1 output.

- New `ViewMeta` specializations for `resize_` and `_unsafe_view` are
  created (see _FunctionalizeFallbackKernel.h_).

- New templates _ViewMetaClasses.{cpp,h}_ are created. They hold the
  declaration and definition of the `ViewMeta` specializations, which
  are automatically generated in the ATen codegen (see _gen.py_).

- New `_functionalization` Python sub-module is created (see
  _Module.cpp_). It serves as namespace for the `ViewMeta`
  specializations and `InverseReturnMode` enum.

- New template _ViewMetaClassesPythonBinding.cpp_ is created. It holds
  the automatically generated Python bindings for the `ViewMeta`
  specialization, which are generated in the torch codegen (see
  _generate_code.py_).

Note that this PR makes use of codegen at 2 different moments:

- ATen codegen (_gen.py_): generates the `ViewMeta` specialized classes.
- Torch codegen (_generate_code.py_): generated the Python bindings for
  them.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

bdhirsh commented on torch/csrc/functionalization/Module.cpp:
```diff
@@ -0,0 +1,33 @@
+void initModule(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+
+  // Create a `torch._C._functionalization` Python module.
```
no action needed: if we're going to make a dedicated `_functionalization` module, we might want to consider moving the existing pybind-ed C helper functions here too (all of the `_functionalize_*` methods in https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/python_torch_functions_manual.cpp#L606)

bdhirsh commented on torchgen/api/functionalization.py:
```diff
@@ -100,7 +110,7 @@ def name(
         assert is_reverse
     if is_reverse:
-        return reverse_name(g.view, include_namespace)
+        return reverse_opname(g.view, include_namespace)
```
any reason for the name change?

ysiraichi replied:
I thought it made more sense to let `name` be the name of the class. Then, `opname` and `reverse_opname` would be the actual function name.

On second thoughts, maybe I can revert this change and have:
- `classname`: name of the class
- `name` and `reverse_name`: actual function name

I guess it would be more consistent with the other apis.

bdhirsh commented on aten/src/ATen/FunctionalizeFallbackKernel.cpp:
```diff
@@ -168,19 +195,8 @@ static const at::Tensor & resize__functionalization(c10::DispatchKeySet dispatch
-    },
-    /*has_symbolic_inputs=*/false
-  );
+  auto view_meta = std::make_shared<at::functionalization::resize__ViewMeta>(
```
hmm, is the `resize_ViewMeta` that you're hand-crafting here any different from the auto-generated one for `as_strided`? (`as_strided_ViewMeta`). If not, we should probably just use the as_strided one.

ysiraichi replied:
This hand-crafted class is not really new. I'm just wrapping the old lambda functions into this class.

That said, as far as I can see:
- `resize_ViewMeta` infers the strides from its size
- `resize_ViewMeta` reverse function always calls `as_strided_scatter`, whereas `as_strided_ViewMeta` may call `as_strided_symint` depending on the `inverse_return_mode`

bdhirsh commented on build.bzl:
```diff
@@ -318,6 +318,10 @@ GENERATED_AUTOGRAD_CPP = [
 ] + _GENERATED_AUTOGRAD_CPP_HEADERS + GENERATED_LAZY_H
 
+GENERATED_PYTHON_CPP = [
+    "torch/csrc/functionalization/generated/ViewMetaClassesPythonBinding.cpp"
```
tbh I would just stick this in the existing `_GENERATED_AUTOGRAD_PYTHON_CPP` to minimize changes to the build system files - that list already includes non-autograd things, like the `torch.*` python bindings.

(changes to the build system scripts have a habit of being more likely to get reverted :p )

bdhirsh reviewed (commented):
overall looks good! Looking forward to the AOTAutograd changes

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
@bdhirsh I have added the missing AOTAutograd changes. At this point, I think this PR is ready for another round of reviews. Here is a brief summary of the latest changes:

**Refactoring:**
- Reverted the function name: `reverse_opname` to `reverse_name`
- Moved _torch/csrc/functionalization/generated/templates/ViewMetaClassesPythonBindings.cpp_ into _aten/src/ATen/templates/ViewMetaClassesPythonBindings.cpp_

**AOTAutograd changes:**
- Moved the method `FunctionalTensorWrapper::apply_view_metas` outside of its class into `impl::apply_view_metas`
- Created a method for accessing the `ViewMeta` sequence outside `FunctionalTensorWrapper`
- Replaced `functional_tensor` with `view_meta_sequence` in `OutputAliasInfo`
- Replaced `FunctionalTensorMetadataEq` by `ViewMetaSequence` class
    - Stores the actual sequence and a `MetadataKey` instance for equality comparison
    - I have to compare with `MetadataKey` because there are cases where we have different `ViewMeta` sequence that end up with the same tensor
- Add `ViewMeta.has_symbolic_inputs` python binding for deciding whether we should use view-replay or not, when generating from base
- Added AOTAutograd cache test

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -218,19 +218,6 @@ def check_cacheable(gm: torch.fx.GraphModule):
-    if config.view_replay_for_aliased_outputs:
-        for info in metadata.output_info:
-            if info.functional_tensor is not None:
-                raise BypassAOTAutogradCache(
```
fyi @jamesjwu

bdhirsh commented on torch/_functorch/_aot_autograd/functional_utils.py:
```diff
@@ -315,28 +310,8 @@ def patch_requires_grad(out):
     return aliased_out
 
 
-def has_same_metadata(t1, t2):
```
nice cleanup

bdhirsh commented on torch/_functorch/_aot_autograd/functional_utils.py:
```diff
@@ -360,25 +335,39 @@ def make(t):
+        self.sequence = _functionalization.get_view_meta_sequence(tensor.elem)
+        self.metadata = MetadataKey.make(tensor)
+
+    def __str__(self) -> str:
```
do you mind including an example of what shows up in TORCH_LOGS="aot" (or tlparse) now, in the case where view-replay happens? It looks like we should get a pretty-printed list of the view metas here (which is great)

bdhirsh reviewed (approved):
nice work!

jamesjwu commented:
This looks great! Could we see if we can change this test to now turn on the config? That way we run all AOTAutograd tests with view replay on and see that it's still cacheable. 

https://github.com/pytorch/pytorch/blob/ec1c3ab3b28143c6e0392352c1c62ae0513ba024/test/functorch/test_aotdispatch.py#L6903

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
CI failure looks unrelated. I will rebase and check if it's gone.

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi commented:
CI is still failing. All of them seem unrelated.

ysiraichi commented:
@pytorchbot merge -i

kit1980 commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

kit1980 commented:
@bdhirsh Can you help to shepherd this diff internally, there are thousands of failed builds, see D68297593

ysiraichi commented:
@bdhirsh Can you help me land this one, again? What did go wrong?

bdhirsh commented:
Taking a look (i'll start looking tonight / tomorrow).

Internal build system probably isn't very happy with the new codegen'd files. I'm going to spend some time trying to appease, but the worst case is that we tweak the PR to not generate any new files and put the generated output into some other existing files (e.g. `templates/CompositeViewCopyKernels.cpp`, which we use for other one-off functionalization codegen today)

ysiraichi commented:
@bdhirsh Let me know if you need me to take any action.

bdhirsh commented:
I'm still battling the internal build issues.

@ysiraichi if you're able to merge as many of the new files as you can into existing files that would be really helpful (if you can't find a good existing file to move the currents contents let me know any we can brainstorm). This other issue I have right now is that... I don't know if we have any precedent for a file like `templates/ViewMetaClassesPythonBinding.cpp`, which is a template but needs to be built as part of building `libtorch_python.so`. Which makes more painful to figure out how to wrangle internally. I'll try to allocate some some more time to look into it this week and report back

ysiraichi commented:
Hi @bdhirsh. Sorry for the very late reply.

> if you're able to merge as many of the new files as you can into existing files that would be really helpful

Sure. Do you mean `ViewMetaClasses.{cpp,h}`? Or all the new files, including the new `functionalization/Module.{cpp,h}`?

> I don't know if we have any precedent for a file like `templates/ViewMetaClassesPythonBinding.cpp`, which is a template but needs to be built as part of building `libtorch_python.so`

Isn't the lazy generated files like that, too?

jamesjwu committed:
Update

[ghstack-poisoned]
